### ISSUE TITLE
Handle numerical errors properly

### ISSUE THREAD

**Issue by mherb** [mherb/kalman#8]

**Comment by LuTaonuwhu** [comment]
Do the numerical errors mean non-positive definite of square root S in SR-UKF and non-positive definite of covariance matrix P in UKF?
for exmaple, here
 S.rankUpdate( U.col(i), -1 );
                
                if( S.info() == Eigen::NumericalIssue )
                {
                    // TODO: handle numerical issues in some sensible way
                    return false;
                }

**Comment by mherb** [comment]
Yes, this is about these kind of errors which usually occur when the covariance matrix is not positive definite.

One approach to handle these would be to force the covariance to be positive definite for instance using the method proposed by [Higham et al.](https://doi.org/10.1016/0024-3795(88)90223-6) (adapted matlab implementation [here](https://de.mathworks.com/matlabcentral/fileexchange/42885-nearestspd)).

Unfortunately I do not have the time right now to implement this, but contributions are of course very welcome.

**Comment by LuTaonuwhu** [comment]
Thanks for your information mherb, I have also notice a solution like [this](https://stats.stackexchange.com/a/6367). I will try to fix it.

**Comment by jwdinius** [comment]
But what if the matrix is negative definite?  There should be a check before even attempting a matrix rescaling; something like:
```
  Eigen::SelfAdjointEigenSolver< Covariance<StateType> > eigenSolver(this->P);
  if (eigenSolver.info() != Eigen::SUCCESS || eigenSolver.eigenvalues().minCoeff() < 0) {
    //! either the solver failed, which could mean the matrix is not symmetric (self-adjoint) OR
    //! there are negative eigenvalues, meaning the covariance matrix is not pos semidefinite
    //! either way, an exception should be raised
  }
```
I like this approach because of the ability to touch the internal filter state and covariance via the `init` and `setCovariance` public methods within the filter base class.  If the caller does something bogus, the implementation could complain about it.
